pkg/chunk/chunk_store.go

```diff
@@ -149,19 +164,59 @@ func (c *store) calculateIndexEntries(userID string, from, through model.Time, c
 	}
 	indexEntriesPerChunk.Observe(float64(len(entries)))
 
+	uncachedEntries, cacheKeys := c.dedupeEntriesFromCache(entries)
+	indexWritesTotal.Add(float64(len(uncachedEntries)))
+
 	// Remove duplicate entries based on tableName:hashValue:rangeValue
 	result := c.storage.NewWriteBatch()
-	for _, entry := range entries {
+	for _, entry := range uncachedEntries {
 		key := fmt.Sprintf("%s:%s:%x", entry.TableName, entry.HashValue, entry.RangeValue)
 		if _, ok := seenIndexEntries[key]; !ok {
 			seenIndexEntries[key] = struct{}{}
 			rowWrites.Observe(entry.HashValue, 1)
 			result.Add(entry.TableName, entry.HashValue, entry.RangeValue, entry.Value)
 		}
 	}
+	c.entryCache.Store(context.Background(), cacheKeys, make([][]byte, len(cacheKeys)))
+
 	return result, nil
 }
 
+func (c *store) dedupeEntriesFromCache(entries []IndexEntry) ([]IndexEntry, []string) {
+	if c.cfg.IndexEntryCacheSize == 0 {
+		return entries, nil
+	}
+
+	// We return the stringified entry and the map from that string->entry
+	// We cannot safely go back from string to entry hence we need the map.
+	keys := make([]string, 0, len(entries))
+	keyMap := make(map[string]IndexEntry, len(entries))
+	for _, entry := range entries {
+		key := strings.Join([]string{
```

Comment:
This is basically repeating the `Sprintf("%s:%s:%x",` from earlier in a different way.  Be more DRY.